src/core/compact_object.h

```diff
@@ -525,14 +482,38 @@ class CompactObj {
   //
   static_assert(sizeof(u_) == 16);
 
-  uint8_t mask_ = 0;
+  union {
+    uint8_t mask_ = 0;
+    struct {
+      uint8_t ref : 1;  // Mark objects that have expiry timestamp assigned.
+      uint8_t expire : 1;
+      uint8_t mc_flag : 1;  // Marks keys that have memcache flags assigned.
+
+      // ascii encoding is not an injective function. it compresses 8 bytes to 7 but also 7 to 7.
+      // therefore, in order to know the original length we introduce 2 flags that
+      // correct the length upon decoding. ASCII1_ENC_BIT rounds down the decoded length,
+      // while ASCII2_ENC_BIT rounds it up. See DecodedLen implementation for more info.
+      uint8_t encoding : 2;
+
+      // IO_PENDING is set when the tiered storage has issued an i/o request to save the value.
+      // It is cleared when the io request finishes or is cancelled.
+      uint8_t io_pending : 1;  // also serves as async-delete for keys.
+      uint8_t sticky : 1;
+
+      // TOUCHED used to determin which items are hot/cold.
+      // by checking if the item was touched from the last time we
+      // reached this item while travering the database to set items as cold.
+      // https://junchengyang.com/publication/nsdi24-SIEVE.pdf
+      uint8_t touched : 1;  // used to mark keys that were accessed.
+    } mask_bits_;
+  };
```

Comment:
What I am not sure about this is that only one field can be active at a time. Either `mask_` or `mask_bits_`. Isn't it very easy to get UB if we don't make sure which field is active ?

Reply:
mask_bits and mask point to the same byte. It's not ub

Reply:
I get it now!